rust/sbd-server/src/cslot.rs

```diff
@@ -0,0 +1,416 @@
+pub struct WeakCSlot(Weak<Mutex<CSlotInner>>);
+
+impl WeakCSlot {
+    pub fn upgrade(&self) -> Option<CSlot> {
+        self.0.upgrade().map(CSlot)
+    }
+}
+
+pub struct CSlot(Arc<Mutex<CSlotInner>>);
+
+impl CSlot {
+    pub fn new(config: Arc<Config>, ip_rate: Arc<ip_rate::IpRate>) -> Self {
+        let count = config.limit_clients as usize;
+        Self(Arc::new_cyclic(|this| {
+            let mut slots = Vec::with_capacity(count);
+            let mut task_list = Vec::with_capacity(count);
+            for _ in 0..count {
+                let (send, recv) = tokio::sync::mpsc::unbounded_channel();
+                slots.push(SlotEntry { send });
+                task_list.push(tokio::task::spawn(top_task(
+                    config.clone(),
+                    ip_rate.clone(),
+                    WeakCSlot(this.clone()),
+                    recv,
+                )));
+            }
+            Mutex::new(CSlotInner {
+                max_count: count,
+                slots,
+                slab: slab::Slab::with_capacity(count),
+                pk_to_index: HashMap::with_capacity(count),
+                ip_to_index: HashMap::with_capacity(count),
+                task_list,
+            })
+        }))
+    }
+
+    pub fn weak(&self) -> WeakCSlot {
+        WeakCSlot(Arc::downgrade(&self.0))
+    }
+
+    fn remove(&self, uniq: u64, index: usize) {
+        let mut lock = self.0.lock().unwrap();
+
+        match lock.slab.get(index) {
+            None => return,
+            Some(s) => {
+                if s.uniq != uniq {
+                    return;
+                }
+            }
+        }
+
+        let _ = lock.slots.get(index).unwrap().send.send(TaskMsg::Close);
+        lock.slab.remove(index);
+        lock.pk_to_index.retain(|_, i| *i != index);
+        lock.ip_to_index.retain(|_, v| {
+            v.retain(|i| *i != index);
+            !v.is_empty()
+        });
+    }
+
+    // oi clippy, this is super straight forward...
+    #[allow(clippy::type_complexity)]
+    fn insert_and_get_rate_send_list(
+        &self,
+        ip: Arc<std::net::Ipv6Addr>,
+        pk: PubKey,
+        ws: Arc<ws::WebSocket<MaybeTlsStream>>,
+    ) -> std::result::Result<
+        Vec<(u64, usize, Weak<ws::WebSocket<MaybeTlsStream>>)>,
+        Arc<ws::WebSocket<MaybeTlsStream>>,
+    > {
+        let mut lock = self.0.lock().unwrap();
+
+        if lock.slab.len() >= lock.max_count {
+            return Err(ws);
+        }
+
+        let weak_ws = Arc::downgrade(&ws);
+
+        let uniq = U.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+
+        let index = lock.slab.insert(SlabEntry {
+            uniq,
+            weak_ws,
+            handshake_complete: false,
+        });
+
+        lock.pk_to_index.insert(pk.clone(), index);
+
+        let rate_send_list = {
+            let list = {
+                // WARN - allocation here!
+                // Also, do we want to limit the max connections from same ip?
```

Comment:
@steveej - I can't read between the typo here. What did you mean to suggest that was *not* public IP?

Reply:
sorry, quite the typo :laughing: i meant public *key* (corrected above as well)

Reply:
Ah, gottcha : ) - Yeah, the main concern here would be an attacker blowing out our memory by opening up a bunch of connections all from the same ip address... since the Vec storing that is currently unbound. Distinguishing by key wouldn't mitigate that since all the connections would have unique keys.